parla/ldevice.py

```diff
@@ -341,3 +342,58 @@ def wrapper(*args, memory, device):
         def wrapper(*args, memory, device):
             return memory(data(*args))
     return wrapper
+
+
+class Coordinator():
+    def __init__(self, latest_view: List):
+        self._latest_view = latest_view
+
+    def __getitem__(self, index):
+        if not isinstance(index, tuple):
+            index = (index,)
+        ret = []
+
+        def traverse(data, prefix, index):
+            if len(index) > 0:
+                i, *rest = index
+                if isinstance(i, slice):
+                    for v in range(i.start or 0, i.stop, i.step or 1):
+                        traverse(data[v], prefix + (v,), rest)
+                elif isinstance(i, Iterable):
+                    for v in i:
+                        traverse(data[v], prefix + (v,), rest)
+                else:
+                    assert isinstance(i, int)
+                    traverse(data[i], prefix + (i,), rest)
+            else:
+                ret.append(clone_here(data) if is_array(data) else data)
+
+        traverse(self._latest_view, (), index)
+        if len(ret) == 1:
+            return ret[0]
+        return ret
+
+    def __setitem__(self, index, value):
+        if not isinstance(index, tuple):
+            index = (index,)
+        ret = []
+
+        def traverse(data, prefix, index):
+            if len(index) > 0:
+                i, *rest = index
+                if isinstance(i, slice):
+                    for v in range(i.start or 0, i.stop, i.step or 1):
+                        traverse(data[v], prefix + (v,), rest)
+                elif isinstance(i, Iterable):
+                    for v in i:
+                        traverse(data[v], prefix + (v,), rest)
+                else:
+                    assert isinstance(i, int)
+                    traverse(data[i], prefix + (i,), rest)
+            else:
+                copy(data, value)
```

Comment:
This code is nearly identical to lines starting at 357. Lets create a function that takes a function to apply at the leaf calls. Here the function will be `lambda data: copy(data, value)`. Above the function will be `lambda data: ret.append(clone_here(data) if is_array(data) else data)`. Prefix is never used in this function. Remove it.

Reply:
You're totally right that I shamelessly stole the traverse from there 🤣. I was actually going to camp it out and mulling over a correct abstraction. Your approach does make sense. The prefix might be useful for future bookkeeping stuff though it's obviously not used here.